selfdrive/ui/qt/widgets/cameraview.cc

```diff
@@ -289,20 +295,23 @@ void CameraViewWidget::vipcThread() {
     }
 
     if (VisionBuf *buf = vipc_client->recv(nullptr, 1000)) {
-      if (!Hardware::EON()) {
-        void *texture_buffer = gl_buffer->map(QOpenGLBuffer::WriteOnly);
-        memcpy(texture_buffer, buf->addr, buf->len);
-        gl_buffer->unmap();
-
-        // copy pixels from PBO to texture object
-        glBindTexture(GL_TEXTURE_2D, texture[buf->idx]->frame_tex);
-        glTexSubImage2D(GL_TEXTURE_2D, 0, 0, 0, buf->width, buf->height, GL_RGB, GL_UNSIGNED_BYTE, 0);
-        glBindTexture(GL_TEXTURE_2D, 0);
-        assert(glGetError() == GL_NO_ERROR);
-        // use glFinish to ensure that the texture has been uploaded.
-        glFinish();
+      {
+        std::lock_guard lk(lock);
+        if (!Hardware::EON()) {
+          void *texture_buffer = gl_buffer->map(QOpenGLBuffer::WriteOnly);
+          memcpy(texture_buffer, buf->addr, buf->len);
+          gl_buffer->unmap();
+
+          // copy pixels from PBO to texture object
+          glBindTexture(GL_TEXTURE_2D, texture[buf->idx]->frame_tex);
+          glTexSubImage2D(GL_TEXTURE_2D, 0, 0, 0, buf->width, buf->height, GL_RGB, GL_UNSIGNED_BYTE, 0);
+          glBindTexture(GL_TEXTURE_2D, 0);
+          assert(glGetError() == GL_NO_ERROR);
+
+          wait_fence.reset(new WaitFence());
```

Comment:
Shouldn't we have a fence per texture?

Reply:
I think one is enough here.  In fact, we don’t actually need multiple textures on c3, one texture is enough. This can be optimized later